src/hooks/useOptions.ts

```diff
@@ -0,0 +1,51 @@
+import React from "react";
+import { ethers } from "ethers";
+import { useClaims } from "./useClaims";
+import { useConnection } from "./useConnection";
+import { contracts } from "../config";
+import { isValidChainId } from "../utils/chainId";
+import { useTransactions } from "./useTransactions";
+
+export function useOptions() {
+  const { account, chainId, signer } = useConnection();
+  const { claims, refetch: refetchClaims } = useClaims();
+  const { addTransaction } = useTransactions();
+  const [error, setError] = React.useState<Error>();
+
+  const claimCallback = React.useCallback(
+    async (claimIdx = 0) => {
+      if (!chainId || !signer || !isValidChainId(chainId)) {
+        return;
+      }
+      // check if we can claim options for this claimIdx.
+      if (!claims || !(claims.length > claimIdx)) {
+        setError(new Error("There are no options to claim for this address."));
+        return;
+      }
+
+      const merkleDistributor = new ethers.Contract(
+        contracts.getMerkleDistributorAddress(chainId),
+        contracts.merkleDistributorABI,
+        signer
+      );
+      const { windowIndex, amount, proof: merkleProof, accountIndex } = claims[
+        claimIdx
+      ];
+
+      merkleDistributor
+        .claim({
+          windowIndex,
+          account,
+          accountIndex,
+          amount,
+          merkleProof,
+        })
+        .then((tx: ethers.ContractTransaction) =>
+          addTransaction(tx, () => refetchClaims())
```

Comment:
its a shame to not be able to use the free promise on the transaction for this work flow, though you have come up with a good solution. but i just feel like this could be simplified since all you need is a global state which assumes a single TX.


```js
// where dispatch uses a global state manager rather than full on tx manager
async function handleTx(transaction,dispatch){
  dispatch('txPending',transaction)
  try{
   await  transaction.wait()
   dispatch('txResolve')
   return transaction
  }catch(err){
    dispatch('txError',err)
    throw err
  }
}

merkleDistributor.claim({... })
  .then((tx: ethers.ContractTransaction) => handleTx(tx,dispatch)
  .catch(setError);
  
```


maybe ultimately this amounts to somewhat of the same thing, but its slightly more generalized in that your provider is a global reducer which could be re-used in other cases. lmk what you think